lc-hax/Scripts/Modules/Possession/PossessionMod.cs

```diff
@@ -85,118 +95,265 @@ internal sealed class PossessionMod : MonoBehaviour {
+        }
+    }
+
+    void KillEnemyAndUnposses() {
+        if (Helper.LocalPlayer is not PlayerControllerB localPlayer) return;
+        if (this.Possession.Enemy is not EnemyAI enemy) return;
+
+        enemy.Kill(localPlayer.actualClientId);
+
+        if (enemy.TryGetComponent(out NetworkObject networkObject)) {
+            networkObject.Despawn(true);
+        }
+
+        this.Unpossess();
     }
 
     // Releases possession of the current enemy
     internal void Unpossess() {
-        this.UnInitCharacterMovement();
         if (this.Possession.Enemy is not EnemyAI enemy) return;
-        if (enemy.agent.Unfake() is NavMeshAgent navMeshAgent) {
-            navMeshAgent.updatePosition = true;
-            navMeshAgent.updateRotation = true;
-
+        if (enemy.agent is NavMeshAgent agent) {
+            agent.updatePosition = true;
+            agent.updateRotation = true;
+            agent.isStopped = false;
             this.UpdateEnemyPosition(enemy);
             _ = enemy.agent.Warp(enemy.transform.position);
         }
 
-        this.SetEnemyColliders(enemy, true);
+        if (this.EnemyControllers.TryGetValue(enemy.GetType(), out IController controller)) {
+            this.HandleEnemyOnUnpossess(controller, enemy);
+        }
+
+        this.IsAIControlled = false;
+        this.MainEntrance = null;
+        this.ResetInteractionCooldowns();
         this.Possession.Clear();
+
+        if (this.CharacterMovement is not null) {
+            Destroy(this.CharacterMovementInstance);
+            this.CharacterMovementInstance = null;
+        }
+    }
+
+    void ToggleAIControl() {
+        if (this.Possession.Enemy?.agent is null) return;
+        if (this.CharacterMovement is null) return;
+        if (this.MousePan is null) return;
+
+        this.IsAIControlled = !this.IsAIControlled;
+        this.SetAIControl(this.IsAIControlled);
+        this.SendPossessionNotifcation($"AI Control: {(this.IsAIControlled ? "Enabled" : "Disabled")}");
+    }
+
+    void SetAIControl(bool enableAI) {
+        if (this.CharacterMovement is not CharacterMovement characterMovement) return;
+        if (this.Possession.Enemy is not EnemyAI enemy) return;
+        if (enemy.agent is not NavMeshAgent agent) return;
+        if (enableAI) {
+            _ = enemy.agent.Warp(enemy.transform.position);
+            enemy.SyncPositionToClients();
+        }
+
+        if (this.NoClipEnabled) {
+            this.NoClipEnabled = false;
+            characterMovement.SetNoClipMode(false);
+        }
+
+        agent.updatePosition = enableAI;
+        agent.updateRotation = enableAI;
+        agent.isStopped = !enableAI;
+        characterMovement.enabled = !enableAI;
+        characterMovement.SetPosition(enemy.transform.position);
+    }
+
+    void ResetInteractionCooldowns() {
+        this.TeleportCooldownRemaining = -DoorInteractionCooldown;
+        this.DoorCooldownRemaining = -TeleportDoorCooldown;
+    }
```

Comment:
Is not a mistake, the negative cooldown is what makes it work

Reply:
```cs
        if (hit.collider.gameObject.TryGetComponent(out DoorLock doorLock) && this.DoorCooldownRemaining <= 0.0f) {
            this.OpenDoorAsEnemy(doorLock);
            this.DoorCooldownRemaining = PossessionMod.DoorInteractionCooldown;
            return;
        }
```

Okay, but it isn't necessary right? I thought it had to be negative or something would break while `0.0f` would be expected and work normally.

Reply:
That cooldown is supposed to go negative, once it hits any negative value it will work  and it will reset

Reply:
`<= 0.0f` means it’s true if it hits `0.0f` right?

Reply:
> `<= 0.0f` means it’s true if it hits `0.0f` right?
0.0 or below, yes